File: source/FuerzaBruta.cpp
#include "FuerzaBruta.h"

#include <iostream>
#include <vector>
#include <limits>


void FB(const std::vector<std::vector<double>>& energia, int fila, std::vector<int> & mejor, std::vector<int> actual, double &mejorEnergia, double energiaActual) {
    int n = energia[0][0];
    int m = energia[0][1];
    
    if(fila == n+1){
        if(energiaActual < mejorEnergia){
            mejor = actual;
            mejorEnergia = energiaActual;
        }
        return;
    }

    if(fila==1){
        for (int col=0; col < m; col++){
            actual.push_back(col);
            FB(energia, fila+1, mejor, actual, mejorEnergia, energiaActual + energia[fila][col]);
            actual.pop_back();
        }
    }

    else{
        int ult_columna = actual.back();

        for (int col = ult_columna - 1; col <= ult_columna + 1; col++) {
            if (col >= 0 && col < m) { // poda de factibilidad
                actual.push_back(col);
                FB(energia, fila + 1, mejor, actual, mejorEnergia, energiaActual + energia[fila][col]);
                actual.pop_back();
            }
        }
    }

}




std::vector<int> encontrarSeamFuerzaBruta(const std::vector<std::vector<double>>& energia) {
    int n = energia[0][0];
    int m = energia[0][1];

    double mejorEnergia = std::numeric_limits<double>::infinity();

    std::vector<int> mejor;
    std::vector<int> actual =  {};
    FB(energia, 1, mejor, actual, mejorEnergia, 0);

    //Para que sea indexado 1
    for(int i=0; i<n; i++){
        mejor[i]++;
    }

    return mejor;
}

File: source/FuerzaBruta.py
def encontrar_seam_backtracking(energia):
    mejor = []

    # inicializo mejor con la columna vertical 0
    for _ in range(len(energia)):
        mejor.append(0)

    actual = []
    return bt(energia, 0, mejor, actual)


def calcular_energia_seam(energia_matriz, seam):
    total = 0.0
    for i in range(len(seam)):
        total += energia_matriz[i][seam[i]]
    return total


def bt(energia, fila, mejor, actual):
    if fila == len(energia):
        if calcular_energia_seam(energia, actual) < calcular_energia_seam(energia, mejor):
            mejor[:] = actual
        return mejor

    if fila == 0:
        for col in range(len(energia[1])):
            actual.append(col)
            bt(energia, fila + 1, mejor, actual)
            actual.pop()

    else:
        # Poda: no seguir costura si energia actual > energia total mejor
        # Poda: recorrer solo casilleros válidos por definición de costura vertical
        if calcular_energia_seam(energia, actual) < calcular_energia_seam(energia, mejor):
            ult_columna = actual[-1]

            actual.append(ult_columna)
            bt(energia, fila + 1, mejor, actual)
            actual.pop()

            if ult_columna > 0:  # primer columna
                actual.append(ult_columna - 1)
                bt(energia, fila + 1, mejor, actual)
                actual.pop()

            if ult_columna < len(energia[1]) - 1:  # ultima columna
                actual.append(ult_columna + 1)
                bt(energia, fila + 1, mejor, actual)
                actual.pop()

    return mejor